test/junit/scala/collection/IteratorTest.scala

```diff
@@ -358,30 +358,48 @@ class IteratorTest {
     assertTrue(hi.hasNext)
   }
   @Test def `flatMap is memory efficient in previous element`(): Unit = {
+    import java.lang.ref._
     // Array.iterator holds onto array reference; by contrast, iterating over List walks tail.
-    // Avoid reaching seq1 through test class.
-    var seq1 = Array("first", "second") // captured, need to set to null
-    var seq11: String = null
+    // Avoid reaching seq1 through test class. Avoid testing Array.iterator.
+    class C extends Iterable[String] {
+      val ss = Array("first", "second")
+
+      def iterator = new Iterator[String] {
+        var i = 0
+
+        def hasNext = i < ss.length
+
+        def next() = if (hasNext) { i += 1; ss(i-1) } else Iterator.empty.next()
+      }
+
+      def apply(i: Int) = ss(i)
+    }
+    val seq1 = new WeakReference(new C)
     val seq2 = List("third")
     val it0: Iterator[Int] = Iterator(1, 2)
     lazy val it: Iterator[String] = it0.flatMap {
-      case 1 => val r = seq1; seq1 = null; seq11 = r(1); r
```

Comment:
This was wrong. Setting `seq1 = null` here actually "breaks" subsequent `check` calls. They are intended to check whether the array is reachable from the iterator, passing `null` was undermining the check. This was not caught because of the buggy `assertThrows`.